Batching rules for complex view functions

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #47227 Batched gradient support for view+inplace operations
* #47365 Add batching rule for torch.clone(tensor, torch.contiguous_format)
* #47364 as_strided batching rule
* #47226 Batching rule for Tensor.new_empty_strided
* #47225 Implement Tensor.new_empty_strided(sizes, strides, *, dtype, device, requires_grad)
* #47223 Batched grad for advanced indexing (index)
* #47189 Fix sum batching rule, add simple clone batching rule
* **#47188 Batching rules for complex view functions**

Includes batching rules for:
- torch.real, torch.imag, torch.view_as_real, and torch.view_as_complex

Test Plan:
- new tests

Differential Revision: [D24741686](https://our.internmc.facebook.com/intern/diff/D24741686)

zou3519 committed:
Batching rules for complex view functions

Includes batching rules for:
- torch.real, torch.imag, torch.view_as_real, and torch.view_as_complex

Test Plan:
- new tests

[ghstack-poisoned]

albanD commented on aten/src/ATen/BatchingRegistrations.cpp:
```diff
@@ -337,13 +337,18 @@ Tensor view_batching_rule(const Tensor& self, IntArrayRef size) {
   return makeBatched(output_physical, BatchDims(old_bdims.begin(), old_bdims.end()));
 }
 
+template <typename F, F Func, typename... ExtraArgs>
```
I am not sure to understand why you need this extra function? How is it different from the one above?

zou3519 replied:
The only thing that is different is th ename. I thought it would be strange if we used "unary_pointwise_batching_rule" to implement batching rules for operations that aren't unary pointwise (e.g., the complex view functions aren't unary pointwise).

zou3519 replied:
But I guess I could just name "unary_pointwise_batching_rule" to "unwrap_and_call"

albanD replied:
Ho ok. I was looking very hard for some subtle template differences.
I guess that using `unwrap_and_call` is better.

albanD commented on aten/src/ATen/BatchingRegistrations.cpp:
```diff
@@ -683,6 +687,17 @@ TORCH_LIBRARY_IMPL(aten, Batched, m) {
+    unwrap_and_call<Tensor (*)(const Tensor&), at::op>);
+  // complex number view operators
+  TRIVIAL_OP(imag)
+  TRIVIAL_OP(conj);
```
conj is not a view for complex Tensor?

zou3519 replied:
Thanks for the catch! I went through this too quickly

albanD commented on aten/src/ATen/BatchingRegistrations.cpp:
```diff
@@ -683,6 +687,17 @@ TORCH_LIBRARY_IMPL(aten, Batched, m) {
+  TRIVIAL_OP(imag)
+  TRIVIAL_OP(conj);
+  TRIVIAL_OP(real);
+  TRIVIAL_OP(view_as_complex);
```
Will this one work fine if the batch dimension is the last one?
This case is explicitly avoided in the tests with the `shape_postfix` argument.

zou3519 replied:
That's a really good point. No, this won't work fine in that case. We'll need a different batching rule here (to move the batch dim to the front).

albanD replied:
But will this movement of dimension be an issue for view operations? As the move might incur a copy?
Should this just be forbidden? (the last dimension has to be the 2 complex values and cannot be a batch dim)

zou3519 replied:
> But will this movement of dimension be an issue for view operations? As the move might incur a copy?

The move won't incur a copy (movement is just a permute).

> Should this just be forbidden? (the last dimension has to be the 2 complex values and cannot be a batch dim)

`view_as_complex` requires that the last dimension has [stride 1 for the view to be valid](https://github.com/pytorch/pytorch/blob/a341a4329a0dd625d890212d72bbd62d3840ba0a/aten/src/ATen/native/ComplexHelper.h#L66) (because we're just viewing [[a0 b0], [a1 b1]] as [a0 b0 a1 b1]. So if the batch dim is at the end of the tensor, we can support `view_as_complex` depending on the strides of the dimension of size 2

```
# This should fail gracefully because the dimension of size 2 doesn't have stride 1.
x = torch.randn(2, B0)
vmap(torch.view_as_complex, in_dim=1)(x)

# This will succeed 
x = torch.randn(B0, 2).t()
vmap(torch.view_as_complex, in_dim=1)(x)

albanD commented on test/test_vmap.py:
```diff
@@ -1210,6 +1210,45 @@ def test_expand_as(self):
+                return torch.randn(shape, dtype=dtype)
+
+            B0, B1 = 7, 11
+            test = self._vmap_view_test
```
nit: this is already defined above

zou3519 committed:
Update on "Batching rules for complex view functions"


Includes batching rules for:
- torch.real, torch.imag, torch.view_as_real, and torch.view_as_complex

Test Plan:
- new tests

[ghstack-poisoned]

albanD reviewed (approved):
lgtm

zou3519 committed:
Update on "Batching rules for complex view functions"


Includes batching rules for:
- torch.real, torch.imag, torch.view_as_real, and torch.view_as_complex

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Batching rules for complex view functions"


Includes batching rules for:
- torch.real, torch.imag, torch.view_as_real, and torch.view_as_complex

Test Plan:
- new tests

[ghstack-poisoned]
